### ISSUE TITLE
EpoxyModelGroup - Static number of models (2) but one of them is has a different layout (view type)

### ISSUE THREAD

**Issue by MohamadAtieh** [airbnb/epoxy#530]
Hey, I am trying to implement card views using `EpoxyModelGroup` with 2 `ViewStub` s. I am making sure that I always pass two models when I add the group in the `buildModels` method. 

However, one of the models is of a different type, and on initial load, the wrong view type is being inflated into the view stub (with data from an existing item). When I scroll away and back, it calls the bind method on the correct type. Sometimes when I scroll down I see it being re-used again with the same data

This is how my build method looks like :

```
override fun buildModels(results: List<Events>) = results.forEachIndexed { index, item ->
        val fieldModel = ImageModel_().apply {
            id("${ImageModel.ITEM_PREFIX}_$index")
            imageUrl(item.url)
            name(item.name)
            date(item.date)
        }

        val infoModel = when (item) {
            is A -> buildA(item)
            is B -> buildB(item)
        }

        // EpoxyModelGroup should always have the same number of models, so we alter the visibility instead
        (infoModel as? AModel)?.run {
            if (!hasData()) {
                hide()
                return@run
            }

            show()
        }

        GroupModel(index, listOf(fieldModel, infoModel)).apply {
            id("${GroupModel.PREFIX}_$index)
            addTo(this@Controller)
        }
    }
```

When I debug this method, `infoModel` is of type `BModel` but what I see on the screen is `AModel` that already exists. The `id` used is `item.id`

Any help would be appreciated.

**Comment by elihart** [comment]
If you have a generated model you should be using it, yes.

You are saying that A and B have models have different layouts? if that's the case this is expected, model groups don't support changing the model's layout because from the group's perspective it is still the same view type id. You need a different view type id to represent the difference - you can set a different layout on the model group to force that

this is related to https://github.com/airbnb/epoxy/issues/289 and the issue with keeping the same number of models. They are both fixable, I just haven't had time for it :(

**Comment by MohamadAtieh** [comment]
Thanks for the response!

- `fieldModel` is fixed to one type / layout
- `infoModel` has a different type / layout
- The group model holding these two models has a single layout (fixed type) & the models list size is always 2

Are you saying that the `infoModel` is expected to be of the same layout? And if we wanted it to be dynamic, the layout / type on the group should be changing?

Btw, I ended up having three `ViewStub` tags and passing in 3 models then altering their visibility based on their data **before** passing them to the group. **_Will this trigger bind on the hidden models?_**

I am not sure if this is OK or having a different group layout is more efficient. 

Thanks!!

**Comment by elihart** [comment]
yes, `infoModel` is expected to be of the same layout. How the model group works is it creates its child views once and then reuses them for all future children, so the models must be the same.

It's fine to hide the models, they won't be bound, just the views will be hidden.
